Set configdata in M2 from database

arnout opened the pull request:
Closes #227

There is no test yet. The easiest way to test this is by adding support for AUTOCONFIG_RENEW to the agent, so that we can re-trigger autoconfiguration in one of the test flows. But that's a bigger change so I'll do it in a separate PR.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -375,6 +375,26 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+{
+    for (uint8_t i = 0; i < radio_basic_caps.operating_classes_info_list_length(); i++) {
+        auto operating_class_info = std::get<1>(radio_basic_caps.operating_classes_info_list(i));
+        if (operating_class_info.operating_class() == bss_info_conf.operating_class) {
```
This will have to be changed after #251 will be merged.

arnout replied:
> This will have to be changed after #251 will be merged.

That's why it's still a draft :-)

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -375,6 +375,26 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
 }
 
+/**
+ * @brief Check if the operating class of @a radio_basic_caps match the operating classes of
```
What is "@a" ?

morantr replied:
@tomereli, try to read the brief in line 379, you will understand what I meant.

arnout replied:
@a is a command to format the following word as code. It is meant to be used for arguments to functions, since those are not auto-formatted (unlike class and function references).

Cfr. http://doxygen.nl/manual/commands.html#cmda

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -375,6 +375,26 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+ */
+static bool has_matching_operating_class(wfa_map::tlvApRadioBasicCapabilities &radio_basic_caps,
+                                         const db::bss_info_conf_t &bss_info_conf)
+{
```
In the past, there was a decision no to add static/helper functions to the son_master_thread since it was big enough, and instead, put them on son_actions.
I don't say it is right, but it is a convention.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -688,11 +716,40 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
+            LOG(INFO) << "Skipping " << bss_info_conf.ssid << " due to encr mismatch";
+            continue;
+        }
+        if (num_bsses >= radio_basic_caps->maximum_number_of_bsss_supported()) {
```
I see the spell of BSS in a plural term is "bsss" (funny IMO), on other places I see BSSes.

arnout replied:
> I see the spell of BSS in a plural term is "bsss" (funny IMO), on other places I see BSSes.

Indeed, `maximum_number_of_bsss_supported` already exists.

morantr commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -688,11 +716,40 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
+        num_bsses++;
+    }
+
+    // If no BSSes (either because none are configured, or because they don't match), tear down.
```
I assumed you meant not to reach here if you called `autoconfig_wsc_add_m2(tlvwscM1, &bss_info_conf)` in line 742, but we will reach here after the previous ranges loop will be finished.

tomereli replied:
So this should be wrapped with
```
if (!num_bsses)
   ....
```

arnout replied:
Oops....

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -618,6 +618,23 @@ bool master_thread::autoconfig_wsc_add_m2(std::shared_ptr<ieee1905_1::tlvWscM1>
     }
 
+    // The MAC address in the config data is tricky... According to "Wi-Fi Simple Configuration
+    // Technical Specification v2.0.4", section 7.2.2 "Validation of Configuration Data" the MAC
```
We should refer to v2.0.6 which is the WSC specification referred to by the EasyMesh Specification, see section 2 
>[5]	Wi-Fi Simple Configuration Technical Specification 2.0.6 (https://www.wi-fi.org/members/certification-programs )

arnout replied:
Meh, I don't have that version of the document and I can't access the WFA site. Can you confirm that the references are still correct?

And then I should try and get version 2.0.6...

tomereli replied:
The references are still correct.

arnout commented:
I will not be able to continue working on this tomorrow, so if anyone wants to take this over on Sunday, feel free.

One thing that is still missing, is updating test_flows.sh with something to actually test this. My idea was to follow the current tests with DEV_RESET_DEFAULT and a sample DEV_SET_CONFIG taken from Table 11, then restart the repeater, then grep on the repeater if it receives the M2 with the correct SSID.

morantr commented on framework/tlvf/yaml/tlvf/WSC/WSC_Attributes.yaml:
```diff
@@ -256,13 +256,13 @@ sWscAttrVendorExtension:
   subelement_value:
     _type: uint8_t
-    _value: TEARDOWN
+    _value: 0x20
```
_value_const: TEARDOWN

arnout replied:
Apparently, _value_const is not supported in struct...

arnout committed:
controller: WSC: create M2 based on bss_info_conf

Get the M2 information from the database and fill in the M2
configuration data based on it.

Match the information from the database with the information from M1:
AL-MAC address, supported authentication and encryption types, and
supported operating classes.

If no BSSes are configured, send the teardown bit.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: WSC: add AL-MAC address to M2 config data

Set the MAC address (bssid) in the M2 config data correctly. Or at
least, set it to something that may be correct.  Note that the
certification test plan does not check this, and that at least one
agent from the certification test (Marvell) ignores it.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented:
@arnout - @morantr is OOO today so if you are ready to push this, you can dismiss his review I assume.
